rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,429 @@
+### Multiplicity of formats
+
+Within nixpkgs alone, several conventions for doc-comments have emerged, see [1], [2] and [3].
+
+Notably, most doc-comments utilize some fraction of the of CommonMark syntax, even if they are not meant to be rendered.
+
+[1]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders/default.nix
+[2]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix
+[3]: https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix
+
+In general, the format for writing documentation strings is **not formally specified**.
+
+Among the formats encountered in the wild, the one used in `nixpkgs/lib` is the only one intended to be rendered as part of an API documentation, via the nixdoc third-party tool, whose syntax has not been standardized.
+
+### Impossible to differentiate from internal comments
+
+The lack of a formal definition of a doc-comment also means there is no reliable way to distinguish them from internal comments, which would result in automatically-produced API documentation which includes the wrong type of comments.
+
+### References to the problems above
+
+> This curated link collection highlights the Nix ecosystem's inconsistencies, a primary focus of this RFC.
+
+#### nixpkgs - comment examples
+
+- [lib/attrsets](https://github.com/NixOS/nixpkgs/blob/master/lib/attrsets.nix)
+- [trivial-builders](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix)
+- [stdenv/mkDerivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/make-derivation.nix)
+- [nixos/lib/make-disk-image](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix)
+
+# Design
+[design]: #detailed-design
+
+In the following we give a comprehensive overview to our decision that we've made. Detailed arguments for and against every decision can be found in the [Decisions](#Decisions) section
+
+## CommonMark
+
+**CommonMark is the content of all doc-comments.**
+
+Adopting CommonMark as the content for all doc-comments brings the benefits of widely accepted and understood documentation format in tech projects, while maintaining profitability and consistency within the Nix ecosystem by aligning with existing [NixOS/RFC-72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md).
+
+## `/** */` is the doc-comment format
+
+The decision to use /** to start a doc-comment ensures a unique distinction from regular comments while still allowing seamless writing without IDE or editor support. This choice not only provides the best developer experience but also minimizes the need for additional tooling overhead.
+
+## Placement
+
+**The placment describes the relationship between doc-comments and the expression they are documenting.**
+
+- Doc-comments are placed before the documentable node. Only `WHITESPACES` are allowed in between. ([Examples](#basic-examples)) 
+    - `WHITESPACES` are: `[\n \r ' ' \t]`.
```

Comment:
Where does this all-caps syntax come from?

It doesn't seem like this RFC should define which characters are considered whitespace.

```suggestion
- Doc-comments are placed before the documentable node. Only whitespace is allowed in between. ([Examples](#basic-examples)) 
```

Would be nice to refer to the lexer or something, but this suggestion is already good enough IMO.

Reply:
Agreed. But i could the lexer could consider comments itself as whitespace too? We didn't want to allow this.

Reply:
The lexer does define it on a separate line, at

https://github.com/NixOS/nix/blob/46028ff76493439921a5a9200d14c015f0a0c025/src/libexpr/lexer.l#L304

So while the parser (ie stage after lexing) can't see comments (or whitespace for that matter) the distinction exists. Whitespace is never understood to include comments, exception perhaps in some language implementations where the authors decide to abbreviate names and not care much about definitions.

No need to worry about this. Whitespace is still just whitespace.